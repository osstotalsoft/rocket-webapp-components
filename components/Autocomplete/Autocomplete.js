/* eslint-disable react/prop-types */
import React, { useCallback } from 'react';
import PropTypes from 'prop-types';
import Select from 'react-select';
import AsyncCreatableSelect from 'react-select/async-creatable';
import AsyncSelect from 'react-select/async';
import { Paper, MenuItem, TextField, ListItem, ListItemIcon, ListItemText, Checkbox, makeStyles, InputAdornment } from '@material-ui/core';
import Typography from '../Typography';
import autoCompleteStyles, { selectedColor } from "./autocompleteStyle"
import { isArray } from 'util';
import { curry, flatten, prop, map, innerJoin, find, propEq, has, all } from 'ramda';
import Search from '@material-ui/icons/Search';
import { useTheme } from '@material-ui/core/styles';

const useStyles = makeStyles(autoCompleteStyles);

function NoOptionsMessage({ selectProps, innerProps, children }) {
  return (
    <Typography
      color={selectProps.typographyProps.color}
      className={selectProps.classes.noOptionsMessage}
      {...innerProps}
    >
      {children}
    </Typography>
  );
}

function inputComponent({ inputRef, ...props }) {
  return <div ref={inputRef} {...props} />;
}

function controlHasValue(selectProps) {
  if (selectProps.inputValue.length > 0) {
    return true;
  }
  if (selectProps.value === null || selectProps.value === undefined) {
    return false;
  }
  if (isArray(selectProps.value) && selectProps.value.length === 0) {
    return false;
  }

  return true;
}

function Control({ selectProps, innerRef, children, innerProps }) {

  const inputProps = {
    inputComponent,
    inputProps: {
      className: selectProps.classes.input,
      inputRef: innerRef,
      children: children,
      ...innerProps,
    },
    endAdornment: (
      <InputAdornment>
        <Search />
      </InputAdornment>)
  }

  if (!selectProps.isSearchable) {
    delete inputProps.endAdornment;
  }

  return (
    <TextField
      fullWidth
      InputProps={inputProps}
      InputLabelProps={{
        shrink: controlHasValue(selectProps),
        style: { fontSize: 'inherit' }
      }}
      value={selectProps.value ? selectProps.getOptionValue(selectProps.value) : null}
      {...selectProps.textFieldProps}
    />
  );
}

function Option({ innerRef, isFocused, isSelected, innerProps, children }) {
  return (
    <MenuItem
      buttonRef={innerRef}
      selected={isFocused}
      component="div"
      style={{
        fontWeight: isSelected ? 500 : 400,
        fontSize: "14px"
      }}
      {...innerProps}
    >
      {typeof children === 'function' ? children() : children}
    </MenuItem>
  );
}

const MultiOption = props => (
  <ListItem dense button selected={props.isSelected} {...props.innerProps}>
    <ListItemIcon>
      <Checkbox
        edge="start"
        checked={props.isSelected}
        tabIndex={-1}
        disableRipple
      />
    </ListItemIcon>
    <ListItemText primary={props.label} />
  </ListItem>

);

function Placeholder({ selectProps, innerProps, children }) {
  return (
    <Typography
      color={selectProps.typographyProps.color}
      className={selectProps.classes.placeholder}
      {...innerProps}
    >
      {children}
    </Typography>
  );
}

function SingleValue({ selectProps, innerProps, children }) {
  const theme = useTheme();

  return (
    <Typography style={selectedColor(theme, selectProps.inputSelectedColor)} className={selectProps.classes.singleValue}{...innerProps}>
      {typeof children === 'function' ? children() : children}
    </Typography>
  );
}

function ValueContainer({ selectProps, children }) {
  return <div className={selectProps.classes.valueContainer}>{children}</div>;
}

function MultiValueDisplay(text) {
  return (
    <span key={text}>
      {text}
    </span>
  );
}

function MultiValueContainer({ selectProps, children }) {

  let elements = Object.assign([], children);
  if (children.length > 0 && children[0] && children[0].length > 1) {
    elements[0] = MultiValueDisplay(`(${children[0].length} selected)`);
  }

  return <div className={selectProps.classes.valueContainer}>{elements}</div>;
}


function MultiValue(props) {
  return (
    <span>{props.children}</span>
  );
}

function Menu({ selectProps, innerProps, children }) {
  return (
    <Paper square className={selectProps.classes.paper} {...innerProps}>
      {children}
    </Paper>
  );
}

function IndicatorSeparator() {
  return (<span />);
}

export const getSimpleValue = (options, value, valueKey, isMultiSelection) => {
  if (isMultiSelection && !Array.isArray(value))
    return null;

  const hasGroups = all(has("options"), options)

  const flattenOptions = hasGroups
    ? flatten(map(prop("options"), options))
    : options

  const result = isMultiSelection
    ? innerJoin((o, v) => o[valueKey] === v, flattenOptions, value)
    : find(propEq(valueKey, value), flattenOptions)
  return result || null;
}

function DropdownIndicator() {
  return (<span />);
}

const formatCreateLabel = curry((createdLabel, text) => createdLabel.concat(text))

function Autocomplete({ options, defaultOptions, loadOptions, onChange,
  creatable, onMenuOpen, value, isMultiSelection, isClearable,
  isSearchable, disabled, simpleValue, label, valueKey, labelKey,
  error, helperText, createdLabel, typographyContentColor, inputSelectedColor, ...other }) {
  const classes = useStyles();

  const handleOnChange = useCallback(inputValue => {
    if (simpleValue && inputValue) {
      if (isMultiSelection) {
        return onChange(inputValue.map(a => a[valueKey]))
      } else {
        return onChange(inputValue[valueKey])
      }
    }

    return onChange(inputValue)
  }, [isMultiSelection, onChange, simpleValue, valueKey])

  const getNewOptionData = useCallback((inputValue, optionLabel) => ({
    [labelKey]: optionLabel,
    [valueKey]: undefined
  }), [labelKey, valueKey])

  const Comp = loadOptions ? creatable ? AsyncCreatableSelect : AsyncSelect : Select;
  const loadOptionsAsync = useCallback(input => {
    return loadOptions(input)
  }, [loadOptions])

  const components = {
    Control,
    Menu,
    MultiValue,
    NoOptionsMessage,
    Option: isMultiSelection ? MultiOption : Option,
    Placeholder,
    SingleValue,
    ValueContainer: isMultiSelection ? MultiValueContainer : ValueContainer,
    IndicatorSeparator,
  }

  if (isSearchable) {
    components.DropdownIndicator = DropdownIndicator;
  }

  return (
    <Comp
      {...other}
      classes={classes}
      styles={classes.selectStyles}
      options={options}
      loadOptions={loadOptionsAsync}
      defaultOptions={defaultOptions}
      components={components}
      closeMenuOnSelect={isMultiSelection ? false : true}
      value={simpleValue ? getSimpleValue(options, value, valueKey, isMultiSelection) : value}
      onChange={handleOnChange}
      onMenuOpen={onMenuOpen}
      textFieldProps={{
        label: label,
        error: error,
        helperText: helperText
      }}
      typographyProps={{
        color: typographyContentColor
      }}
      multiValueRemove={false}
      creatable={creatable}
      hideSelectedOptions={false}
      placeholder={null}
      isMulti={isMultiSelection}
      isClearable={isClearable}
      isSearchable={isSearchable}
      isDisabled={disabled}
      getOptionLabel={prop(labelKey)}
      getOptionValue={prop(valueKey)}
      formatCreateLabel={formatCreateLabel(createdLabel)}
      getNewOptionData={getNewOptionData}
      inputSelectedColor={inputSelectedColor}
    />
  )
}

Autocomplete.defaultProps = {
  options: [],
  isMultiSelection: false,
  isClearable: false,
  isSearchable: false,
  disabled: false,
  simpleValue: false,
  valueKey: "id",
  labelKey: "name",
  error: false,
  value: null,
  creatable: false,
  typographyContentColor: "textSecondary"
};

Autocomplete.propTypes = {
  /**
  * The array of options from which the client can select a value.
  */
  options: PropTypes.array,
  /**
  * Function that returns a promise, which is the set of options to be used once the promise resolves.
  */
  loadOptions: PropTypes.func,
  /**
  * The selected value from list of options.
  */
  value: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array,
    PropTypes.number,
    PropTypes.string,
    PropTypes.bool
  ]),
  /**
  * Handle change events on the select.
  */
  onChange: PropTypes.func.isRequired,
  /**
  * Handle the menu opening.
  */
  onMenuOpen: PropTypes.func,
  /**
  * If true, the user can select multiple values from list.
  */
  isMultiSelection: PropTypes.bool,
  /**
  * If true, the user can clear the selected value.
  */
  isClearable: PropTypes.bool,
  /**
  * If true, the search functionality is enabled.
  */
  isSearchable: PropTypes.bool,
  /**
  * If true, the Select is disabled.
  */
  disabled: PropTypes.bool,
  /**
  * @TODO
  */
  simpleValue: PropTypes.bool,
  /**
  * Label to be displayed in the heading component.
  */
  label: PropTypes.string,
  /**
  * The key of values from options.
  * @TODO
  */
  valueKey: PropTypes.string,
  /**
  * @TODO
  */
  labelKey: PropTypes.string,
  /**
  * If true, the helper text is displayed when an error pops up.
  */
  error: PropTypes.bool,
  /**
  * The value of label when a new option is added.
  */
  createdLabel: PropTypes.string,
  /**
  * The default set of options to show before the user starts searching. When set to true, the results for loadOptions('') will be autoloaded.
  */
  defaultOptions: PropTypes.oneOfType([PropTypes.bool, PropTypes.array]),
  /**
  * The color of both the text displayed when there are no options and placeholder. It supports those theme colors that make sense for this component.
  */
  typographyContentColor: PropTypes.oneOf([
    'initial',
    'inherit',
    'primary',
    'secondary',
    'textPrimary',
    'textSecondary',
    'error'
  ]),
  /**
  * The color of selected input.
  */
  inputSelectedColor: PropTypes.string
};

export default Autocomplete;