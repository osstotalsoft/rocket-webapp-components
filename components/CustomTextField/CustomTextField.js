/* eslint-disable react/prop-types */
import React, { useState, useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import NumberFormat from "react-number-format";
import { makeStyles, TextField } from "@material-ui/core";
import textFieldStyle from "./textFieldStyle";
import { useDebounce } from "use-debounce";
import { curry } from "ramda";
import { emptyFunction, emptyString } from "../../utils/constants";

const useStyles = makeStyles(textFieldStyle);

const isAllowed = values =>
  !(values.floatValue && values.floatValue.toString().includes("e"));

function NumberFormatCustom(props) {
  const {
    inputRef,
    onChange,
    decimalScale,
    fixedDecimalScale,
    thousandSeparator,
    language,
    ...other
  } = props;
  let formatter = new Intl.NumberFormat(language);

  var thousandSep = thousandSeparator
    ? formatter.format(1111).replace(/1/g, "")
    : thousandSeparator;
  var decimalSeparator = formatter.format(1.1).replace(/1/g, "");
  const valueIsNumericString =
    !!props.value &&
    typeof props.value === "string" &&
    !Number.isNaN(Number(props.value));

  const handleValueChange = useCallback(values => onChange(values.floatValue), [
    onChange
  ]);

  return (
    <NumberFormat
      {...other}
      getInputRef={inputRef}
      style={{ textAlign: "right" }}
      onValueChange={handleValueChange}
      isAllowed={isAllowed}
      decimalScale={decimalScale}
      fixedDecimalScale={fixedDecimalScale}
      thousandSeparator={thousandSep}
      decimalSeparator={decimalSeparator}
      isNumericString={valueIsNumericString}
      //format = {customFormat}
    />
  );
}

NumberFormatCustom.propTypes = {
  inputRef: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  decimalScale: PropTypes.number.isRequired,
  fixedDecimalScale: PropTypes.bool,
  thousandSeparator: PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
};

NumberFormatCustom.defaultProps = {
  decimalScale: 2,
  fixedDecimalScale: true,
  thousandSeparator: true
};

// TODO: this is for backwards compatibility (onTextBoxChange) ... remove onTextBoxChange and keep only value
const setValueToDesiredFormat = curry((isNumeric, value) =>
  isNumeric ? value : { target: { value } }
);
const getValueFromDesiredFormat = curry((isNumeric, e) =>
  isNumeric ? e : e?.target.value
);

function CustomTextField({
  isNumeric,
  customInputProps,
  inputProps,
  endAdornment,
  startAdornment,
  InputLabelProps,
  className,
  value,
  onChange,
  debounceBy,
  ...rest
}) {
  const classes = useStyles();

  // props applied to the Input element
  const customMuiInput = isNumeric
    ? {
        inputComponent: NumberFormatCustom,
        endAdornment,
        startAdornment,
        ...customInputProps
      }
    : {
        endAdornment,
        startAdornment,
        ...customInputProps
      };

  // attributes applied to the input element
  const customReactInput = {
    ...inputProps,
    className: `${classes.input} ${
      inputProps?.className ? inputProps.className : emptyString
    }`
  };

  const [localValue, setLocalValue] = useState(value);
  const [debouncedValue] = useDebounce(localValue, debounceBy);

  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    debouncedValue !== value &&
      onChange(setValueToDesiredFormat(isNumeric)(debouncedValue));
  }, [debouncedValue]);
  // eslint-disable-next-line react-hooks/exhaustive-deps
  useEffect(() => {
    value !== localValue && setLocalValue(value);
  }, [value]);

  const handleChange = useCallback(
    e => setLocalValue(getValueFromDesiredFormat(isNumeric)(e)),
    [isNumeric]
  );

  return (
    <TextField
      onChange={handleChange}
      value={localValue}
      {...rest}
      className={classes.textField + " " + className}
      InputProps={customMuiInput}
      inputProps={customReactInput}
      InputLabelProps={{
        className: classes.label,
        ...InputLabelProps
      }}
    />
  );
}

CustomTextField.defaultProps = {
  isNumeric: false,
  debounceBy: 0,
  onChange: emptyFunction
};

CustomTextField.propTypes = {
  /**
   * Override or extend the styles applied to the component.
   */
  className: PropTypes.string,
  /**
   * If true, the input will accept only numeric values.
   */
  isNumeric: PropTypes.bool,
  /**
   * Other properties you can provide to the Input component.
   */
  customInputProps: PropTypes.object,
  /**
   * Attributes applied to the input element.
   */
  inputProps: PropTypes.object,
  /**
   * End adornment of componenent. (Usually an InputAdornment from material-ui)
   */
  endAdornment: PropTypes.any,
  /**
   * Start adornment of componenent. (Usually an InputAdornment from material-ui)
   */
  startAdornment: PropTypes.any,
  /**
   * Props applied to the InputLabel element.
   */
  InputLabelProps: PropTypes.object,
  /**
   * The value of the `input` element, required for a controlled component.
   */
  value: PropTypes.any,
  /**
   * Callback fired when the value is changed.
   *
   * @param {object} event The event source of the callback.
   * You can pull out the new value by accessing `event.target.value` (string).
   */
  onChange: PropTypes.func,
  /**
   * The delay of debouncing.
   */
  debounceBy: PropTypes.number,
  /**
   * The current language, preferably taken from the i18next (i18.language) or another internationalization library
   */
  language: PropTypes.string
};

export default CustomTextField;
