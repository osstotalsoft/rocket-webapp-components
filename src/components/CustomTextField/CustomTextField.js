/* eslint-disable react/prop-types */
import React, { useState, useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import NumberFormat from "react-number-format";
import { makeStyles, TextField, deprecatedPropType } from "@material-ui/core";
import textFieldStyle from "./textFieldStyle";
import { useDebounce } from "use-debounce";
import { curry, is, head } from "ramda";
import { emptyFunction, emptyString } from "../../utils/constants";

const useStyles = makeStyles(textFieldStyle);

const isAllowed = values =>
  !(values.floatValue && values.floatValue.toString().includes("e"));

function NumberFormatCustom(props) {
  const {
    inputRef,
    onChange,
    decimalScale,
    fixedDecimalScale,
    thousandSeparator,
    language,
    ...other
  } = props;
  let formatter = new Intl.NumberFormat(language);

  var thousandSep = thousandSeparator
    ? formatter.format(1111).replace(/1/g, "")
    : thousandSeparator;
  var decimalSeparator = formatter.format(1.1).replace(/1/g, "");
  const valueIsNumericString =
    !!props.value &&
    typeof props.value === "string" &&
    !Number.isNaN(Number(props.value));

  const handleValueChange = useCallback(values => onChange(values.floatValue), [
    onChange
  ]);

  return (
    <NumberFormat
      {...other}
      getInputRef={inputRef}
      style={{ textAlign: "right" }}
      onValueChange={handleValueChange}
      isAllowed={isAllowed}
      decimalScale={decimalScale}
      fixedDecimalScale={fixedDecimalScale}
      thousandSeparator={thousandSep}
      decimalSeparator={decimalSeparator}
      isNumericString={valueIsNumericString}
      //format = {customFormat}
    />
  );
}

NumberFormatCustom.propTypes = {
  inputRef: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  decimalScale: PropTypes.number.isRequired,
  fixedDecimalScale: PropTypes.bool,
  thousandSeparator: PropTypes.oneOfType([PropTypes.string, PropTypes.bool])
};

NumberFormatCustom.defaultProps = {
  decimalScale: 2,
  fixedDecimalScale: true,
  thousandSeparator: true
};

// TODO: this is for backwards compatibility (onTextBoxChange) ... remove onTextBoxChange and keep only value
const setValueToDesiredFormat = curry((isNumeric, value) =>
  isNumeric ? value : { target: { value } }
);
const getValueFromDesiredFormat = curry((isNumeric, e) =>
  isNumeric ? e : e?.target.value
);

function CustomTextField({
  isNumeric,
  customInputProps,
  inputProps,
  InputProps,
  endAdornment,
  startAdornment,
  InputLabelProps,
  className,
  value,
  onChange,
  debounceBy,
  helperText,
  ...rest
}) {
  const classes = useStyles();

  // props applied to the Input element
  const customMuiInput = isNumeric
    ? {
        inputComponent: NumberFormatCustom,
        endAdornment,
        startAdornment
      }
    : {
        endAdornment,
        startAdornment,
        ...InputProps
      };

  // attributes applied to the input element
  const customReactInput = {
    ...inputProps,
    className: `${classes.input} ${
      inputProps?.className ? inputProps.className : emptyString
    }`
  };

  const [localValue, setLocalValue] = useState(value);
  const [debouncedValue] = useDebounce(localValue, debounceBy);

  useEffect(() => {
    debouncedValue !== value &&
      onChange(setValueToDesiredFormat(isNumeric)(debouncedValue));
  }, [debouncedValue]); // eslint-disable-line react-hooks/exhaustive-deps

  useEffect(() => {
    value !== localValue && setLocalValue(value);
  }, [value]); // eslint-disable-line react-hooks/exhaustive-deps

  const handleChange = useCallback(
    e => setLocalValue(getValueFromDesiredFormat(isNumeric)(e)),
    [isNumeric]
  );

  return (
    <TextField
      onChange={handleChange}
      value={localValue}
      helperText={is(Array, helperText) ? head(helperText) : helperText}
      {...rest}
      className={className ? `${classes.textField} ${className}` : classes.textField}
      InputProps={customMuiInput}
      inputProps={customReactInput}
      InputLabelProps={{
        className: classes.label,
        ...InputLabelProps
      }}
    />
  );
}

CustomTextField.defaultProps = {
  isNumeric: false,
  debounceBy: 0,
  onChange: emptyFunction
};

CustomTextField.propTypes = {
  /**
   * Override or extend the styles applied to the component.
   */
  className: PropTypes.string,
  /**
   * If true, the input will accept only numeric values.
   */
  isNumeric: PropTypes.bool,
  /**
   * Other properties you can provide to the component.
   */
  customInputProps: deprecatedPropType(PropTypes.object, "Use `inputProps` instead."),
  /**
   * Attributes applied to the input element.
   * For the numeric input, you can provide properties like thousandSeparator, decimalScale and allowNegative.
   */
  inputProps: PropTypes.object,
  /**
   * Other properties you can provide to the Input component.
   */
  InputProps: PropTypes.object,
  /**
   * End adornment of componenent. (Usually an InputAdornment from material-ui)
   */
  endAdornment: PropTypes.any,
  /**
   * Start adornment of componenent. (Usually an InputAdornment from material-ui)
   */
  startAdornment: PropTypes.any,
  /**
   * Props applied to the InputLabel element.
   */
  InputLabelProps: PropTypes.object,
  /**
   * The value of the `input` element, required for a controlled component.
   */
  value: PropTypes.any,
  /**
   * The content of the helper under the input.
   */
  helperText: PropTypes.node,
  /**
   * Callback fired when the value is changed.
   *
   * @param {object} event The event source of the callback.
   * You can pull out the new value by accessing `event.target.value` (string).
   */
  onChange: PropTypes.func,
  /**
   * The delay of debouncing.
   */
  debounceBy: PropTypes.number,
  /**
   * The current language, preferably taken from the i18next (i18.language) or another internationalization library
   */
  language: PropTypes.string
};

export default CustomTextField;
