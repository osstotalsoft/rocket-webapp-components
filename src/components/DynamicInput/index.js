import DynamicInput from "./DynamicInput";
export default DynamicInput;