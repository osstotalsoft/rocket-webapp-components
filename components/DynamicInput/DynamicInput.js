import React, { useState, useCallback } from "react";
import PropTypes from "prop-types";
import { FormControlLabel, Switch } from "@material-ui/core";
import AutoComplete from "../Autocomplete";
import CustomTextField from "../CustomTextField";
import { emptyArray, emptyString } from "../../utils/constants";

function DynamicInput({ controlType, onChange, loadOptions, value, DefaultComp, defaultCompProps, ...other }) {
  const [comboOptions, setComboOptions] = useState(emptyArray);

  const comboOpen = useCallback(async () => {
    if (comboOptions.length === 0) {
      const options = await loadOptions();

      setComboOptions(
        options.map(option => {
          return { option: option.value };
        })
      );
    }
  }, [comboOptions.length, loadOptions])

  const onInputChange = useCallback(event => {
    onChange(event.target.value);
  }, [onChange])

  const onSelectionChange = useCallback(inputValue => {
    onChange(inputValue.option);
  }, [onChange])

  const onPropertyChange = useCallback(inputValue => {
    onChange(inputValue);
  }, [onChange])

  const onCheckboxPropertyChange = useCallback((_, inputValue) => {
    onChange(inputValue);
  }, [onChange])

  const integer = <CustomTextField
    isNumeric={true}
    value={value || emptyString}
    onChange={onPropertyChange}
    fullWidth
    {...other}
  />;
  const numericField = <CustomTextField
    onChange={onPropertyChange}
    isNumeric={true}
    fullWidth
    inputProps={{ decimalScale: 0, thousandSeparator: emptyString }}
    value={value || emptyString}
    {...other} />;
  const comboField = <AutoComplete {...other}
    fullWidth
    labelKey="option"
    valueKey="option"
    value={{ option: value }}
    onChange={onSelectionChange}
    options={comboOptions}
    onMenuOpen={comboOpen}
  />;
  const boolField = <FormControlLabel
    control={
      <Switch
        {...other}
        checked={value}
        onChange={onCheckboxPropertyChange} />
    }
  />
  const stringField = <CustomTextField onChange={onInputChange}
    fullWidth
    value={value || emptyString}
    {...other} />;

  const defaultField = <DefaultComp
    value={value || ""}
    onChange={onInputChange}
    {...defaultCompProps}
  />

  switch (controlType) {
    case 'INT': return integer;
    case 'NUM': return numericField;
    case 'STR': return stringField;
    case 'CLST': return comboField;
    case 'VCLST': return comboField;
    case 'BOOL': return boolField;
    default:
      return defaultField;
  }
}

DynamicInput.defaultProps = {
  DefaultComp: CustomTextField,
  defaultCompProps: { fullWidth: true, disabled: true }
};

DynamicInput.propTypes = {
  /**
  * The title of button.
  */
  controlType: PropTypes.oneOf([
    'INT',
    'NUM',
    'STR',
    'CLST',
    'VCLST',
    'BOOL'
  ]),
  /**
  * Callback fired when the value is changed.
  *
  * @param {object} event The event source of the callback.
  * You can pull out the new value by accessing `event.target.value` (string).
  */
  onChange: PropTypes.func.isRequired,
  /**
  * Load the options.
  */
  loadOptions: PropTypes.func,
  /**
  * The value of the `input` element, required for a controlled component.
  */
  value: PropTypes.any,
  /**
  * The default component.
  */
  DefaultComp: PropTypes.func,
  /**
  * The default props of the default component.
  */
  defaultCompProps: PropTypes.any
};

export default DynamicInput;