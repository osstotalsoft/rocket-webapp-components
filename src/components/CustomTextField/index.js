import CustomTextField from "./CustomTextField";
export default CustomTextField;