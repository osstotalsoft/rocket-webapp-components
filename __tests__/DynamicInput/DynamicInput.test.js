import React from 'react';
import DynamicInput from '../../components/DynamicInput/DynamicInput';
import { mount } from 'enzyme';
import CustomTextField from '../../components/CustomTextField';
import Autocomplete from '../../components/Autocomplete';
import { Switch } from "@material-ui/core";

jest.mock('@material-ui/core/styles', () => ({
    ...jest.requireActual('@material-ui/core/styles'),
    useTheme: () => {
        return {
            autoComplete: {
                valueContainer: {
                    display: 'flex',
                    flexWrap: 'wrap',
                    flex: 1,
                    alignItems: 'center',
                    paddingBottom: '5px'
                },
                input: {
                    fontFamily: 'Source Sans Pro !important',
                    fontSize: '14px !important',
                    fontWeight: 'normal !important',
                    fontStretch: 'normal !important',
                    fontStyle: 'normal !important',
                    lineHeight: '1.29 !important',
                    color: '#a8a8a8 !important',
                },
                selectedInputColor: "#000"
            }
        }
    }
}));

describe('DynamicInput', () => {
    it('if controlType is INT then the input is a CustomTextField without inputProps associated', () => {

        const onChangeMock = jest.fn();

        const wrapper = mount(<DynamicInput
            id={'testId'}
            controlType={"INT"}
            onChange={onChangeMock}
            value={"test"}
            disabled={false}
        />);

        expect(wrapper.find(CustomTextField)).toHaveLength(1);
        expect(wrapper.find(CustomTextField).props().inputProps).toBeUndefined();
    });

    it('if controlType is INT then the input is a CustomTextField with inputProps associated', () => {

        const onChangeMock = jest.fn();

        const wrapper = mount(<DynamicInput
            id={'testId'}
            controlType={"NUM"}
            onChange={onChangeMock}
            value={"test"}
            disabled={false}
        />);

        expect(wrapper.find(CustomTextField)).toHaveLength(1);
        expect(wrapper.find(CustomTextField).props().inputProps).not.toBeUndefined();
    });

    it('if controlType is NUM then the input is a CustomTextField with inputProps associated', () => {

        const onChangeMock = jest.fn();

        const wrapper = mount(<DynamicInput
            id={'testId'}
            controlType={"NUM"}
            onChange={onChangeMock}
            value={"test"}
            disabled={false}
        />);

        expect(wrapper.find(CustomTextField)).toHaveLength(1);
        expect(wrapper.find(CustomTextField).props().inputProps).not.toBeUndefined();
    });

    it('if controlType is CLST then the input is an Autocomplete', () => {

        const onChangeMock = jest.fn();

        const wrapper = mount(<DynamicInput
            id={'testId'}
            controlType={"CLST"}
            onChange={onChangeMock}
            value={"test"}
            disabled={false}
        />);

        expect(wrapper.find(Autocomplete)).toHaveLength(1);
    });

    it('if controlType is BOOL then the input is a Switch', () => {

        const onChangeMock = jest.fn();

        const wrapper = mount(<DynamicInput
            id={'testId'}
            controlType={"BOOL"}
            onChange={onChangeMock}
            value={true}
            disabled={false}
        />);

        expect(wrapper.find(Switch)).toHaveLength(1);
    });

    it('should rended a default component if controlType does not match any other case label', () => {

        const onChangeMock = jest.fn();

        const wrapper = mount(<DynamicInput
            id={'testId'}
            onChange={onChangeMock}
            value={true}
            disabled={false}
            DefaultComp={CustomTextField}
        />);

        expect(wrapper.find(CustomTextField)).toHaveLength(1);
    });
});