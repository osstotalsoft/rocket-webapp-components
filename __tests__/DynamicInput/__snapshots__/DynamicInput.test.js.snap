// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DynamicInput snapshot 1`] = `
<div
  className="MuiFormControl-root MuiTextField-root undefined undefined MuiFormControl-fullWidth"
>
  <label
    className="MuiFormLabel-root MuiInputLabel-root makeStyles-label-1 MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
    data-shrink={true}
    htmlFor="testId"
    id="testId-label"
  >
    DynamicInput - NumberTextField
  </label>
  <div
    className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl"
    onClick={[Function]}
  >
    <input
      aria-invalid={false}
      autoFocus={false}
      className="MuiInputBase-input MuiInput-input makeStyles-input-2"
      disabled={false}
      id="testId"
      inputMode="numeric"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onMouseUp={[Function]}
      required={false}
      style={
        Object {
          "textAlign": "right",
        }
      }
      type="text"
      value=""
    />
  </div>
</div>
`;
